src/time.rs

```diff
@@ -0,0 +1,278 @@
+//! Wall clock time
+//!
+//! Time is fixed to nanosecond precision, with user supplied resolution.
+
+/// 61 bit nanosecond wall clock time
+///
+/// Note that the upper 3 bits are reserved for internal use.
+///
+/// This can represent approximately 73 years.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+#[repr(transparent)]
+pub struct Nanoseconds(u64);
+
+/// Nanosecond time resolution
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+#[repr(transparent)]
+pub struct NanosecondResolution(pub u32);
+
+/// Nanosecond high bits
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+#[repr(transparent)]
+pub struct NanosecondsHighBits(pub u32);
+
+/// Nanosecond low bits
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+#[repr(transparent)]
+pub struct NanosecondsLowBits(pub u32);
+
+/// Wall clock identifier, indicates the time domain
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+#[repr(transparent)]
+pub struct WallClockId(pub u16);
+
+impl Nanoseconds {
+    /// The largest permissible backing time value
+    pub const MAX: Self = Nanoseconds(
+        0b0001_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111,
+    );
+
+    /// One seconds worth of nanoseconds
+    pub const SECOND: Self = Nanoseconds(1_000_000_000);
+    /// One milliseconds worth of nanoseconds
+    pub const MILLISECOND: Self = Nanoseconds(1_000_000);
+    /// One microseconds worth of nanoseconds
+    pub const MICROSECOND: Self = Nanoseconds(1_000);
+
+    /// Create a zero value time
+    pub const fn zero() -> Self {
```

Comment:
May consider making this a `const ZERO: Nanoseconds` as well for ease of comparison.